core/store/src/cold_storage.rs

```diff
@@ -59,35 +77,49 @@ pub fn update_cold_db(
     hot_store: &Store,
     shard_layout: &ShardLayout,
     height: &BlockHeight,
+    num_threads: usize,
 ) -> io::Result<bool> {
     let _span = tracing::debug_span!(target: "cold_store", "update cold db", height = height);
     let _timer = metrics::COLD_COPY_DURATION.start_timer();
 
-    let mut store_with_cache = StoreWithCache { store: hot_store, cache: StoreCache::new() };
-
-    if store_with_cache.get(DBCol::BlockHeight, &height.to_le_bytes())?.is_none() {
+    if hot_store.get_for_cold(DBCol::BlockHeight, &height.to_le_bytes())?.is_none() {
         return Ok(false);
     }
 
     let height_key = height.to_le_bytes();
-    let block_hash_vec = store_with_cache.get_or_err(DBCol::BlockHeight, &height_key)?;
+    let block_hash_vec = hot_store.get_or_err_for_cold(DBCol::BlockHeight, &height_key)?;
     let block_hash_key = block_hash_vec.as_slice();
 
     let key_type_to_keys =
-        get_keys_from_store(&mut store_with_cache, shard_layout, &height_key, block_hash_key)?;
-    for col in DBCol::iter() {
-        if !col.is_cold() {
-            continue;
-        }
-
-        if col == DBCol::State {
-            copy_state_from_store(shard_layout, block_hash_key, cold_db, &mut store_with_cache)?;
-            continue;
-        }
-
-        let keys = combine_keys(&key_type_to_keys, &col.key_type());
-        copy_from_store(cold_db, &mut store_with_cache, col, keys)?;
-    }
+        get_keys_from_store(&hot_store, shard_layout, &height_key, block_hash_key)?;
+    let cold_columns = DBCol::iter().filter(|col| col.is_cold()).collect::<Vec<DBCol>>();
+
+    rayon::ThreadPoolBuilder::new()
+        .num_threads(num_threads)
+        .build()
+        .map_err(|_| io::Error::new(io::ErrorKind::Other, "Failed to create rayon pool"))?
+        .install(|| {
+            cold_columns
+                .into_par_iter()
+                .map(|col: DBCol| -> io::Result<()> {
+                    if col == DBCol::State {
+                        copy_state_from_store(shard_layout, block_hash_key, cold_db, &hot_store)
+                    } else {
+                        let keys = combine_keys(&key_type_to_keys, &col.key_type());
+                        copy_from_store(cold_db, &hot_store, col, keys)
+                    }
+                })
```

Comment:
Really cool! 

I would multiline this a bit to make it more readable but up to you if you like it this way it's fine.